website/blog/2022-10-31-simple-1.mdx

```diff
@@ -0,0 +1,108 @@
+---
+slug: simple-functions-1
+title: "Simple Functions: Introduction and Basic Optimizations"
+authors: [lsakka]
+tags: [tech-blog, simple-functions]
+---
+
+_This blogpost is part of a series of blog posts that discuss different features and optimizations of the simple function interface in Velox_.
+
+
+## Introduction to Simple Functions
+
+Scalar functions are one of the most used extension points in Velox. Since Velox is a vectorized engine, by nature functions are <a href="https://facebookincubator.github.io/velox/develop/scalar-functions.html#vector-functions">"vector functions"</a> that consume Velox vectors (batches of data) and produce vectors. Velox allows users to write functions as vector functions or as single-row operations "simple functions" that are converted to vector functions using template expansion through <a href="https://github.com/facebookincubator/velox/blob/main/velox/expression/SimpleFunctionAdapter.h">SimpleFunctionAdapter</a>.
+
+<figure>
+<img src="/img/simple1_1.png"/>
+</figure>
+
+Writing functions as vector functions directly gives the user `complete` control over the function implementations and optimizations, however it comes with some cost that can be summarized in three points:
+
+- **Complexity** : Requires an understanding of Velox vectorized data representation and encodings, which requires additional work for our customers, specially those without DB background. And writing optimized vector functions requires even deeper understanding.
+- **Repetition** : Involves repeated efforts and code; in each function vectors are read in the same way, same set of optimizations are applied ... etc.
+- **Reliability** : More code means more bugs, especially in such a complex context.
+
+Writing functions through the simple interface mitigates the previously mentioned drawbacks, and significantly simplifies the function authoring process. For example, to add the function plus the user only needs to implement the PlusFunction struct shown in the graph above , which is then expanded using template expansion to a vector function.
+
+However, the simple function interface does not give the user full control and has its own limitations, for example the function map_keys can be implemented in O(1) as a vector function by moving the keys vector; this is not possible to express as a simple function.
+
```

Comment:
it would be nice to discuss at some point that there's usually no perf regression with using simple functions. Maybe could even show that nice chart we have with some examples?

Reply:
I have those for complex types do not have benchmarks for simple primitive operations the next blog post talking about complex types will have a bunch of those:
https://github.com/facebookincubator/velox/pull/3349

I can mention that for arithmetic operations for example, simple and vector implementations have the same performance.

Reply:
updated las paragpraph:

In the past year, we have been working on several improvements to the simple function interface on both the `expressivity` and `performance` axes that we will discuss in this series of notes. 
In this blog post, we will talk about some of the general optimizations that we have in the adapter, the optimizations discussed in this post make the performance of most simple function that operates on primitive types matches their optimized vector function counter implementations. In the next blog post, we will discuss complex types in simple functions.